js.go

```diff
@@ -1005,11 +1022,61 @@ func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, isSync
 	// If we are pull based go ahead and fire off the first request to populate.
 	if isPullMode {
 		sub.jsi.pull = o.pull
+	} else if hasFC || hasHeartbeats {
+		// Start goroutine to handle flow control messages for a push consumer.
+		sub.jsi.fch = make(chan *Msg, 64)
+		go sub.processControlFlow()
```

Comment:
Why do we need this?

Reply:
I moved the control messages handling to be done before delivering the message at `processMsg` and added an extra goroutine that would be handling the flow control messages from a subscription instead: https://github.com/nats-io/nats.go/pull/692/files#diff-1d2fb9af677137c0182213770500dd5625a51ba7286301cb11d0197b21e0f948R2533

Reply:
Anyway to avoid having another Go routine?

Reply:
I removed the goroutine and now responding a flow control message happens as soon as we receive the message, we have to do the processing of control messages there to be able to support ChanSubscribe:

- https://github.com/nats-io/nats.go/pull/692/files#diff-1d2fb9af677137c0182213770500dd5625a51ba7286301cb11d0197b21e0f948R2543
- https://github.com/nats-io/nats.go/pull/692/files#diff-176a2342c55218b70f165bed6a7731dc83d901f9513382f3c98bbd3d08ce0678R1047